### ISSUE TITLE
Support X-Forwarded-Proto header for httpToHttpsRedirect

### ISSUE THREAD

**Issue by sfllaw** [sourcegraph/issues#73]
- Issue type: Bug report
- Sourcegraph version: 2.7.6
- OS Version: N/A
- Docker version: N/A

We run Sourcegraph, with only port 80 exposed, behind Amazon's Application Load Balancer which terminates our TLS connection. If we set [`httpToHttpsRedirect = true`](https://about.sourcegraph.com/docs/config/settings/#httptohttpsredirect-boolean), browsers going to the load balancer error out with `ERR_TOO_MANY_REDIRECTS`.

It looks like Sourcegraph's `frontend` doesn't understand the [X-Forwarded-Proto](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto) header, which is set to `https` if the load balancer or reverse proxy has already done the HTTPS termination: [Redirect HTTP to HTTPS Using ELB](https://aws.amazon.com/premiumsupport/knowledge-center/redirect-http-https-elb/)

`frontend` should understand X-Forwarded-Proto (and the standardized [Forwarded](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) header) which reverse-proxying webservers use to indicate that they are terminating TLS. This would provide compatibility with not just Amazon's load balancers, but also [Squid](http://www.squid-cache.org/Doc/config/forwarded_for/), [Apache](https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#x-headers), [nginx](https://www.nginx.com/resources/wiki/start/topics/examples/forwarded/), _etc_.

Of course, this behavior must be configurable because you don't want a bare Sourcegraph instance to get spoofed by a fake header when it isn't behind a load balancer or reverse proxy. Perhaps `httpToHttpsRedirect` should take a string value of `on`, `off`, or `load-balanced`?

#### Steps to reproduce:
1. Set up a Sourcegraph instance to listen on port 80 only.
2. Set up a load balancer or reverse proxy that terminates TLS and forwards to Sourcegraph on port 80
3. Turn on `httpToHttpsRedirect` in the configuration
4. Go to https://load-balancer and receive an `ERR_TOO_MANY_REDIRECTS`

**Comment by keegancsmith** [comment]
@sfllaw thanks for the detailed report. You mention that you would want to be able to turn off interpreting `X-Forwarded-Scheme` header  due to security reasons. I'm failing to see how it would affect Sourcegraph? Do you mind giving an example?

MITM isn't an issue, since ignoring X-Forwarded-Scheme won't stop whatever MITM agent might do.

The only possibility I can think of is maybe your TLS proxy does authentication via client certs, but just proxies through port 80 and assumes that will always redirect by the application. But I'd think in a situation like that, one shouldn't assume the above.

FYI we never adjust our behaviour based on the request scheme.

**Comment by sfllaw** [comment]
@keegancsmith There are two cases that the server needs to handle, but it doesn't know enough about how it's configured to autodetect the correct behaviour:

**1. Sourcegraph is running directly on the public Internet**

Sourcegraph, by default, should ignore `X-Forwarded-Proto` on port 80, when it comes to respecting `httpToHttpsRedirect`. This is because any HTTP client can send an `X-Forwarded-Proto` header. One creative way to do this might be to fire an XmlHttpRequest in a browser on a completely different site, in order to listen to the session cookie.

**2. Sourcegraph is running behind a load balancer or reverse proxy**

The load balancer or reverse proxy would clobber `X-Forwarded-Proto`, so it is safe to trust it completely. This is why setting `httpToHttpsRedirect = load-balanced` is a good solution, because it's self-documenting.

**Comment by keegancsmith** [comment]
Sorry for asking more questions, but I prefer to minimize the state space of our config so would just like to understand a concrete example.

> Sourcegraph is running directly on the public Internet

I may just be missing something, but how does an XmlHttpRequest on HTTP vs HTTPS change what could be leaked to a random site? Sure if someone could listen in on your connection and convince you to visit a certain page we could potentially leak something. If that is what you meant and you think that is reasonable to protect against I'll add it. But outside of that, the CORS and CSRF protection should stop random requests.

**Comment by sfllaw** [comment]
@keegancsmith Yes, someone could listen to your connection and convince you to visit the HTTP site, when the admin had set `httpToHttpsRedirect = true`. If you blindly trust `X-Forwarded-Proto`, then they can ignore that directive and keep the connection downgraded to HTTP, which is the opposite of what you'd want.

**Comment by keegancsmith** [comment]
@sfllaw I am adjusting our implementation such that we only interpret `X-Forwarded-Proto` if `httpToHttpsRedirect = "load-balanced"`. Thanks for the advice.

**Comment by sfllaw** [comment]
Fixed in 15680_2018-05-18_2d07482.
